src/CartesianProduct.jl

```diff
@@ -1,132 +1,251 @@
 import Base.*
 
-export CartesianProduct, CartesianProductArray, is_contained
+export CartesianProduct,
+       CartesianProductArray,
+       is_contained
 
 """
-    CartesianProduct <: LazySet
+    CartesianProduct{S1<:LazySet,S2<:LazySet} <: LazySet
 
-Type that represents the cartesian product.
+Type that represents a Cartesian product of two convex sets.
 
 ### Fields
 
-- `X` -- convex set
-- `Y` -- another convex set
+- `X` -- first convex set
+- `Y` -- second convex set
 
 ### Notes
 
-The Cartesian product of three elements is done recursively. See also
-`CartesianProductArray` for implementation of a cartesian product of
-many sets without recursion, using instead an array.
-"""
-struct CartesianProduct{T1<:LazySet,T2<:LazySet} <: LazySet
-    X::T1
-    Y::T2
-    CartesianProduct{T1,T2}(X::T1, Y::T2) where {T1<:LazySet,T2<:LazySet} = new(X, Y)
-    CartesianProduct{T}(Xarr::Vector{T}) where {T<:LazySet} = length(Xarr) == 0 ?
-            VoidSet(1) : (length(Xarr) == 1 ? Xarr[1] :
-            new{T,T}(Xarr[1], CartesianProduct{T}(Xarr[2:length(Xarr)])))
-            # NOTE: use array type instead of element type (bit of a mess otherwise)
+The Cartesian product of three elements is obtained recursively.
+See also `CartesianProductArray` for an implementation of a Cartesian product of
+many sets without recursion, instead using an array.
+
+- `CartesianProduct{S1<:LazySet,S2<:LazySet}` -- default constructor
+
+- `CartesianProduct(Xarr::Vector{S}) where {S<:LazySet}`
+  -- constructor from an array of convex sets
+"""
+struct CartesianProduct{S1<:LazySet,S2<:LazySet} <: LazySet
+    X::S1
+    Y::S2
 end
-CartesianProduct(X::T1, Y::T2) where {T1<:LazySet,T2<:LazySet} = CartesianProduct{T1,T2}(X, Y)
-CartesianProduct(Xarr::Vector{T}) where {T<:LazySet} = CartesianProduct{T}(Xarr)
+CartesianProduct(Xarr::Vector{S}) where {S<:LazySet} =
+    (length(Xarr) == 0
+        ? VoidSet(1)
+        : length(Xarr) == 1
+            ? Xarr[1]
+            : length(Xarr) == 2
+                ? CartesianProduct(Xarr[1], Xarr[2])
+                : CartesianProduct(Xarr[1],
+                                   CartesianProduct(Xarr[2:length(Xarr)])))
+
+"""
```

Comment:
there is something strange going on -- from this point on, this file gets commented out.

Reply:
Strange, I cannot find any mistake and Julia does not complain either.
Maybe a bug in Github?